test/performance/odp_random.c

```diff
@@ -0,0 +1,376 @@
+/* Copyright (c) 2021, Nokia
+ *
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <getopt.h>
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+
+#define KB 1024ull
+#define MB (1024ull * 1024ull)
+
+/* Command line options */
+typedef struct {
+	int num_threads;
+	uint32_t size;
+	uint32_t rounds;
+} options_t;
+
+static options_t options;
+static const options_t options_def = {
+	.num_threads = 1,
+	.size = 256,
+	.rounds = 100000,
+};
+
+static void print_usage(void)
+{
+	printf("\n"
+	       "random data performance test\n"
+	       "\n"
+	       "Usage: odp_random [options]\n"
+	       "\n"
+	       "  -t, --threads Number of worker threads (default %u)\n"
+	       "  -s, --size    Size of buffer in bytes (default %u)\n"
+	       "  -r, --rounds  Number of test rounds (default %u)\n"
+	       "                Rounded down to nearest multiple of 8\n"
+	       "  -h, --help    This help\n"
+	       "\n",
+	       options_def.num_threads, options_def.size, options_def.rounds);
+}
+
+static int parse_options(int argc, char *argv[])
+{
+	int opt;
+	int long_index;
+	int ret = 0;
+
+	static const struct option longopts[] = {
+		{ "threads", required_argument, NULL, 't' },
+		{ "size", required_argument, NULL, 's' },
+		{ "rounds", required_argument, NULL, 'r' },
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+
+	static const char *shortopts = "+t:s:r:h";
+
+	options = options_def;
+
+	while (1) {
+		opt = getopt_long(argc, argv, shortopts, longopts, &long_index);
+
+		if (opt == -1)
+			break;
+
+		switch (opt) {
+		case 't':
+			options.num_threads = atol(optarg);
+			break;
+		case 's':
+			options.size = atol(optarg);
+			break;
+		case 'r':
+			options.rounds = atol(optarg);
+			break;
+		case 'h':
+			/* fall through */
+		default:
+			print_usage();
+			ret = -1;
+			break;
+		}
+	}
+
+	if (options.size < 1) {
+		ODPH_ERR("Invalid size: %" PRIu32 "\n", options.size);
+		return -1;
+	}
+
+	return ret;
+}
+
+const char *shm_name = "odp_random_test";
+
+typedef struct test_shm_t {
+	odp_barrier_t barrier;
+	odp_random_kind_t type;
+	uint64_t nsec[ODP_THREAD_COUNT_MAX];
+} test_shm_t;
+
+static test_shm_t *shm_lookup(void)
+{
+	test_shm_t *shm = NULL;
+	odp_shm_t shm_hdl = odp_shm_lookup(shm_name);
+
+	if (shm_hdl != ODP_SHM_INVALID)
+		shm = (test_shm_t *)odp_shm_addr(shm_hdl);
+
+	return shm;
+}
+
+static int test_random(void *p)
+{
+	(void)p;
+
+	uint8_t *buf, *data;
+	const unsigned long page = 4 * KB;
+	odp_time_t start;
+	uint64_t nsec;
+	test_shm_t *shm = shm_lookup();
+
+	if (!shm) {
+		ODPH_ERR("Failed to look up shm %s\n", shm_name);
+		exit(EXIT_FAILURE);
+	}
+
+	/* One extra page for alignment, another one for offset. */
```

Comment:
> another one for offset

What does this mean? Why is the second extra page needed?

Reply:
It's not, this is just a copy-paste error. Fixed in v4.